### PR TITLE
pm: Add option to have synchronous runtime power management

### PR DESCRIPTION
**Pull request by nordic-krch**
Current solution assumes that power management operations may be blocking, asynchronous and take unknown amount of time. Due to this assumption runtime PM API cannot be effectively used from the interrupt context. Zephyr has few driver APIs which can be used from an interrupt context (UART async and polling,  rtio work also targets asynchronous ops) and now use of runtime PM is limited in those cases. Given that In many cases suspending or resuming of a device is limited to just a few register writes having whole infrastructure with semaphore, event and work queue is like using sledgehammer to crack a nut.

Patch introduces a new type of PM device - synchronous PM. If device is specified as capable of synchronous PM operations then device runtime getting and putting is executed in the critical  section. In that case, runtime API can be used from an interrupt context. Additionally, this approach reduces RAM needed for PM device (104 -> 20 bytes of RAM on ARM Cortex-M) because PM data structure is limited and does not contain work,event and semaphore.

If driver has synchronous PM operations (like only changing pinctrl state) then PM device can be defined as
```
PM_DEVICE_SYNC_DEFINE(test_driver, test_driver_action); // instead of PM_DEVICE_DEFINE

PM_DEVICE_SYNC_DT_DEFINE(UARTE(idx), uarte_nrfx_pm_action) // instead of PM_DEVICE_DT_DEFINE
```

Runtime API stays the same. Synchronous PM device property is detected inside the API.

### CODE REVIEW

**Comment by JordanYates** [comment]
The feature makes a lot of sense, I'm not a fan of the name however.

The current basic operations are described as synchronous and asynchronous depending on whether the `pm_device_runtime_*` call blocks or not. This now introduces a second synchronous meaning, whether the transition is run directly by the call or deferred out to a workqueue. Maybe borrow terminology from ISR and call it [direct](https://docs.zephyrproject.org/latest/kernel/services/interrupts.html#defining-a-direct-isr)?

**Comment by nordic-krch** [comment]
@JordanYates thanks for the comment. I knew already that name is wrong and was hoping to get some good suggestions in review. `direct` sounds good to me.

File: include/zephyr/pm/device.h [hunk 1]
```diff
@@ -231,21 +302,26 @@ struct pm_device {
...
-	Z_PM_DEVICE_DEFINE_SLOT(dev_id);				\
-	static struct pm_device Z_PM_DEVICE_NAME(dev_id) =		\
-	Z_PM_DEVICE_INIT(Z_PM_DEVICE_NAME(dev_id), node_id,		\
-			 pm_action_cb)
+#define Z_PM_DEVICE_DEFINE(node_id, dev_id, pm_action_cb)	\
+	Z_PM_DEVICE_DEFINE_SLOT(dev_id);			\
+	static struct pm_device_rt Z_PM_DEVICE_NAME(dev_id) =	\
+	Z_PM_DEVICE_INIT(Z_PM_DEVICE_NAME(dev_id), node_id, pm_action_cb)
+
+#define Z_PM_DEVICE_SYNC_DEFINE(node_id, dev_id, pm_action_cb)				\
```
Comment (ceolin) [inline include/zephyr/pm/device.h:310]: Can't we avoid additional macros ... 
Maybe  something around ...
```
#define PM_DEVICE_DEFINE(dev_id, pm_action_cb, ...) \
	COND_CODE_1(Z_IS_EMPTY(__VA_ARGS__), \
	(Z_PM_DEVICE_DEFINE(DT_INVALID_NODE, dev_id, pm_action_cb)), \
	(Z_PM_DEVICE_SYNC_DEFINE(DT_INVALID_NODE, dev_id, pm_action_cb)))
```

**Review by ceolin — CHANGES_REQUESTED** [review]
Thanks for the pr. I think it makes sense, but we need better nomenclature / clarification here.

The current implementation supports synchronous calls and can be called from ISR. So, this pr is not introducing a new feature. What it is doing is saving resources for devices with no complex power management. It is reducing the size of pm struct for these devices and avoid additional cycles when they are suspending / resuming.

But this makes things confuse because no `PM_DEVICE_FLAG_SYNC_LOCKED_CAPABLE` devices can also be called from ISR ...

What I think can be a good route is, devices without this flag can not longer be called from ISR (anyways it is risky right now since we don't have any hint if the `pm->cb` call will block, and renaming `SYNC_LOCKED`  with something more explicit like `ISR_SAFE` or so.

**Comment by nordic-krch** [comment]
@ceolin 
> The current implementation supports synchronous calls and can be called from ISR. So, this pr is not introducing a new feature.

Current implementation supports putting from isr (async put) but getting may fail if we interrupt another getting operation (it will fail on taking locking semaphore). If operation is not reliable then we cannot fully claim that it is supported.

File: include/zephyr/pm/device.h [hunk 2]
```diff
@@ -122,32 +124,69 @@ typedef bool (*pm_device_action_failed_cb_t)(const struct device *dev,
...
 };
 
+/**
+ * @brief Runtime PM info for device with synchronous PM.
+ *
+ * Synchronous PM can be used with devices which suspend and resume operations can
+ * be performed in the critical section as they are short and non-blocking.
+ * Runtime PM API can be used from any context in that case.
+ */
+struct pm_device_rt_sync {
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply include/zephyr/pm/device.h:177]: isn't `pm_base` to generic. maybe `pm_device_base`?

**Comment by nordic-krch** [comment]
Would be good to agree on the name then. DIRECT(proposed by @JordanYates)  ISR_SAFE (proposed by @ceolin ). I don't have strong opinions.

File: include/zephyr/pm/device.h (hunk 1 above)
Comment (nordic-krch) [reply include/zephyr/pm/device.h]: Sounds good to me if we extend API macro `PM_DEVICE_DEFINE(node, dev, cb, ...)`  I am not sure what would be the additional argument. Should it be something like `PM_DEVICE_DEFINE(..., PM_DEVICE_TYPE_<name>)` where flag can indicate generic device or the new one and for backward compatibility no argument would mean the generic runtime pm?

**Comment by ceolin** [comment]
> @ceolin
> 
> > The current implementation supports synchronous calls and can be called from ISR. So, this pr is not introducing a new feature.
> 
> Current implementation supports putting from isr (async put) but getting may fail if we interrupt another getting operation (it will fail on taking locking semaphore). If operation is not reliable then we cannot fully claim that it is supported.

That is how you take it, the way I see is that the feature is supported and that is its behavior when called from ISR. But, that is not really the point I wanted to make. I agree with you that this  proposed approach is better and gives up something more realiable / deterministic.

Note that naming is not the only thing I've pointed, I think that if we have this explicit option, we should prohibit devices not using it from be get/put from ISR.

File: include/zephyr/pm/device.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (ceolin) [reply include/zephyr/pm/device.h:177]: yep ... `pm_device_base` sounds better.

```
struct pm_device_base;  
struct pm_device;  
struct pm_device_isr;  

```

File: include/zephyr/pm/device.h (hunk 1 above)
Comment (nordic-krch) [reply include/zephyr/pm/device.h]: Extended that was. Optional flag can be provided to indicate ISR safe type.
`PM_DEVICE_DEVICE(dev, cb)` standard device
`PM_DEVICE_DEVICE(dev, cb, 0)` standard device
`PM_DEVICE_DEVICE(dev, cb, PM_DEVICE_ISR_SAFE)` ISR safe device

`PM_DEVICE_ISR_SAFE` is just defined as 1.

**Comment by carlescufi** [comment]
@ceolin please take another look

**Review by ceolin — APPROVED** [review]
I think some internal apis can have their name changed, from sync to isr to avoid confusion, but this can be done later since does not change the behavior or public interface.

Thanks a lot for putting this together !

**Review by carlescufi — APPROVED** [review]

File: include/zephyr/device.h [hunk 3]
```diff
@@ -900,7 +906,7 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
 		.state = (state_),                                             \
 		.data = (data_),                                               \
 		IF_ENABLED(CONFIG_DEVICE_DEPS, (.deps = (deps_),)) /**/        \
-		IF_ENABLED(CONFIG_PM_DEVICE, (.pm = (pm_),)) /**/              \
+		IF_ENABLED(CONFIG_PM_DEVICE, (.pm_base = (pm_),)) /**/         \
```
Comment (marc-hb) [inline include/zephyr/device.h:909]: Looks like this broke gcc 4.2.0 again.

Seems very similar to https://github.com/zephyrproject-rtos/zephyr/pull/68118

```
 -c zephyr/soc/xtensa/intel_adsp/common/mem_window.c
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:62: error: unknown field ‘pm_base’ specified in initializer
cc1: warnings being treated as errors
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:62: warning: missing braces around initializer
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:62: warning: (near initialization for ‘__device_dts_ord_66.<anonymous>’)
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:65: error: unknown field ‘pm_base’ specified in initializer
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:65: warning: missing braces around initializer
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:65: warning: (near initialization for ‘__device_dts_ord_67.<anonymous>’)
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:68: error: unknown field ‘pm_base’ specified in initializer
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:68: warning: missing braces around initializer
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:68: warning: (near initialization for ‘__device_dts_ord_68.<anonymous>’)
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:71: error: unknown field ‘pm_base’ specified in initializer
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:71: warning: missing braces around initializer
zephyr/soc/xtensa/intel_adsp/common/mem_window.c:71: warning: (near initialization for ‘__device_dts_ord_69.<anonymous>’)


```

Comment (marc-hb) [reply include/zephyr/device.h:909]: Fix submitted:
- #69270

**Comment by JordanYates** [comment]
I think this was missed due to the synchronous/asynchronous naming problems, but this PR changed the behaviour of `pm_device_put_async` for devices with `PM_DEVICE_ISR_SAFE` in a way that doesn't conform with the API (#88217).

**Pull request merged**